tfhe/src/core_crypto/gpu/vec.rs

```diff
@@ -25,25 +29,189 @@ pub struct CudaVec<T: Numeric> {
+    /// as soon as synchronization is required
+    pub unsafe fn copy_from_gpu_async(&mut self, src: &Self, stream: &CudaStream)
+    where
+        T: Numeric,
+    {
+        assert!(self.len() >= src.len());
+        let size = src.len() * std::mem::size_of::<T>();
+        // We check that src is not empty to avoid invalid pointers
+        if size > 0 {
+            cuda_memcpy_async_gpu_to_gpu(
+                self.as_mut_c_ptr(),
+                src.as_c_ptr(),
+                size as u64,
+                stream.as_c_ptr(),
+            );
+        }
+    }
+
+    /// Copies data between two `CudaVec`, selecting a range of `src` as target
+    ///
+    /// # Safety
+    ///
+    /// - [CudaStream::synchronize] __must__ be called after the copy
+    /// as soon as synchronization is required
+    pub unsafe fn copy_src_range_gpu_to_gpu_async<R>(
+        &mut self,
+        range: R,
+        src: &Self,
+        stream: &CudaStream,
+    ) where
+        R: std::ops::RangeBounds<usize>,
+        T: Numeric,
+    {
+        let (start, end) = src.range_bounds_to_start_end(range).into_inner();
+        // size is > 0 thanks to this check
+        if end < start {
+            return;
+        }
+        assert!(end < src.len());
+        assert!(end - start < self.len());
+
+        let src_ptr = src.as_c_ptr().add(start * std::mem::size_of::<T>());
+        let size = (end - start + 1) * std::mem::size_of::<T>();
+        cuda_memcpy_async_gpu_to_gpu(self.as_mut_c_ptr(), src_ptr, size as u64, stream.as_c_ptr());
+    }
+
+    /// Copies data between two `CudaVec`, selecting a range of `self` as target
+    ///
+    /// # Safety
+    ///
+    /// - [CudaStream::synchronize] __must__ be called after the copy
+    /// as soon as synchronization is required
+    pub unsafe fn copy_self_range_gpu_to_gpu_async<R>(
+        &mut self,
+        range: R,
+        src: &Self,
+        stream: &CudaStream,
+    ) where
+        R: std::ops::RangeBounds<usize>,
+        T: Numeric,
+    {
+        let (start, end) = self.range_bounds_to_start_end(range).into_inner();
+        // size is > 0 thanks to this check
+        if end < start {
+            return;
+        }
+        assert!(end < self.len());
+        assert!(end - start < src.len());
+
+        let dest_ptr = self.as_mut_c_ptr().add(start * std::mem::size_of::<T>());
+        let size = (end - start + 1) * std::mem::size_of::<T>();
+        cuda_memcpy_async_gpu_to_gpu(dest_ptr, src.as_c_ptr(), size as u64, stream.as_c_ptr());
+    }
+
+    /// Copies data from `CudaVec` into slice
+    ///
+    /// # Safety
+    ///
+    /// - [CudaStream::synchronize] __must__ be called as soon as synchronization is
+    /// required
+    pub unsafe fn copy_to_cpu_async(&self, dest: &mut [T], stream: &CudaStream)
+    where
+        T: Numeric,
+    {
+        let src_size = self.len() * std::mem::size_of::<T>();
+        assert!(std::mem::size_of_val(dest) >= src_size);
```

Comment:
Shouldn't we be checking the lenghts here ? and non size_of

Reply:
Indeed, thanks for catching it.

Reply:
Now I'm wondering if I should not check that they have the same len(), instead of `dest.len() >= src.len()`

Reply:
@tmontaigu I've checked for equality of len() for `copy_to_gpu_async`,  `copy_from_gpu_async` and `copy_to_cpu_async`.

Reply:
Hmm we're actually doing copies where the target buffer has a size greater than the source. For now it's simpler to keep the original checks.